### PR TITLE
Opt to enable linkerd proxy shutdown from lifecycle preStop

### PR DESCRIPTION
**Pull request by thedodd**
#### What this PR does / why we need it
Added a new `linkerd` object to values & schema. If `enabled=true` & `shutdownAfterCheckpoint=true`, then the STS pod's `lifecycle.preStop` hook will be configured to issue a shutdown call to the local linkerd-proxy.

This is entirely optional. If a deployment is not using linkerd, then this option should be left in the default disabled state.

#### Which issue this PR fixes
N/A

#### Special notes for your reviewer
N/A

#### Checklist

- [x] Chart Version bumped
- [x] [CLA signed](https://cla-assistant.io/timescale/helm-charts)

### CODE REVIEW

**Review by graveland — APPROVED** [review]
lgtm!

File: charts/timescaledb-single/scripts/prestop.sql [hunk 1]
```diff
@@ -60,3 +60,7 @@ CHECKPOINT;
     SELECT 'We are a primary: Successfully invoked checkpoints, now issuing a switchover.';
     \! curl -s http://localhost:8008/switchover -XPOST -d '{"leader": "$(hostname)"}'
 \endif
+
+\if :linkerdShutdown
+    \! curl -s -m 5 -X POST http://localhost:4191/shutdown
```
Comment (alexeyklyukin) [inline charts/timescaledb-single/scripts/prestop.sql:65]: Will the linkerd shutdown disconnect the pod from the outside connections, in particular, S3 and replicas? We still need to archive the shutdown checkpoint, so it is only safe to cut the pod from the network when PostgreSQL is fully stopped.

File: charts/timescaledb-single/docs/admin-guide.md [hunk 2]
```diff
@@ -36,6 +36,8 @@ The following table lists the configurable parameters of the TimescaleDB Helm ch
 | `image.pullPolicy`                | The pull policy                             | `IfNotPresent`                                      |
 | `image.repository`                | The image to pull                           | `timescale/timescaledb-ha`                       |
 | `image.tag`                       | The version of the image to pull            | `pg13-ts2.1-latest`
+| `linkerd.enabled`                 | Enabled limited linkerd support.            | `false`
```
Comment (alexeyklyukin) [inline charts/timescaledb-single/docs/admin-guide.md:39]: seems we have two parameters to toggle a single action (shutdown checkpoint)?

File: charts/timescaledb-single/scripts/prestop.sql (hunk 1 above)
Comment (thedodd) [reply charts/timescaledb-single/scripts/prestop.sql:65]: CHECKPOINT should not return until that has been performed, as long as archival is configured. Else, is it even safe to call `/switchover` a few lines above?

Comment (thedodd) [reply charts/timescaledb-single/scripts/prestop.sql:65]: Well, Patroni would/should handle ensuring that it is safe to perform a switchover; so I suppose the primary concern is about ensuring that the checkpoint is archived.

I'll add a line to check `SELECT * FROM pg_stat_archiver;` to ensure that we everything is finished before we issue the linkerd shutdown.

Comment (thedodd) [reply charts/timescaledb-single/scripts/prestop.sql:65]: Ok, posting a quick follow-up here. The call to /shutdown terminates active connections pretty much immediately. And it does NOT appear as though `CHECKPOINT` does any waiting at all for an archival process to finish.

I'm thinking that our best bet right now is likely to check the status of the archiver: `pgbackrest info --output=json | jq -r '.[0].archive'`. Maybe cross-referencing that info with pg_stat_archiver and such.

Comment (thedodd) [reply charts/timescaledb-single/scripts/prestop.sql:65]: Ok, I think there is actually a much more simple solution here. Let me put the commit together real quick.

File: charts/timescaledb-single/docs/admin-guide.md (hunk 2 above)
Comment (thedodd) [reply charts/timescaledb-single/docs/admin-guide.md:39]: Fair enough, will refactor.

File: charts/timescaledb-single/scripts/prestop.sql (hunk 1 above)
Comment (thedodd) [reply charts/timescaledb-single/scripts/prestop.sql]: @alexeyklyukin I'm going to close this thread, as there are no longer changes in this specific location. When you have the time, please let me know what you think of the new solution.

File: charts/timescaledb-single/docs/admin-guide.md (hunk 2 above)
Comment (thedodd) [reply charts/timescaledb-single/docs/admin-guide.md]: Done.

**Comment by thedodd** [comment]
Ok, quick heads-up that I've reverted this to draft from, as I need to do some additional experimentation here.

File: charts/timescaledb-single/scripts/liveness_probe.sh [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+# However, there is such a thing as zombie processes ... so let's just check anyway.
+if pgrep "pgbackrest"; then
+    echo "pgbackrest is still kicking, exit 0"
+    exit 0
+fi
+
+echo "PG is not ready, patroni process is gone, no pgbackrest operations detected, this thing is dead"
+
+# First param indicates that we should attempt to shutdown the local linkerd-proxy on exit.
+if [[ $1 == "1" ]]; then
```
Comment (ncalibey) [inline charts/timescaledb-single/scripts/liveness_probe.sh:45]: Nit: `"$1"`. Could even `(( $1 == 1 ))` -- but whichever

**Review by ncalibey — APPROVED** [review]

**Comment by ncalibey** [comment]
Maybe also do `set -euo pipeline` in the script just to be safe?

**Comment by thedodd** [comment]
Closing this for now as we will be exploring a better way to handle these conditions outside of the chart.

**Pull request closed without merging**